Documentation Cleanup

keefertaylor opened the pull request:
- Update README to show contributors, link to contributions and link to top level documentation at Xpring-SDK
- Fix build status badge
- Add contributing documentation (Text taken from "Contributor Covenant Code of Conduct", recommended by GitHub)
- Add Code of Conduct

intelliot commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+Thanks for considering a contribution to the [Xpring SDK](https://github.com/xpring-eng/xpring-sdk)!
+
+We're thrilled your interested and your help is greatly appreciated. Contributing is a great way to learn more about how the [XRP Ledger](https://xrpl.org) and [Interledger Protocol (ILP)](https://interledger.org/). We are happy to review your pull requests. To make the process as smooth as possible, please read this document and follow the stated guidelines.
```
`your interested` -> `you're interested`

`to learn more about how the` -> `to learn about the`

keefertaylor replied:
Embarrassing. Fixed :)

intelliot commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+TODO: Diagram of Xpring SDK
+
+This library is made up of [protocol buffers](https://developers.google.com/protocol-buffers) which form common model objects for the Xpring SDK. There are also [gRPC service definitions](https://grpc.io), which form a network interace for the Xpring SDK.
```
`interace` -> `interface`

keefertaylor replied:
Done

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+This library is made up of [protocol buffers](https://developers.google.com/protocol-buffers) which form common model objects for the Xpring SDK. There are also [gRPC service definitions](https://grpc.io), which form a network interace for the Xpring SDK.
+
+This library is consumed by both the [Xpring Common JavaScript](https://github.com/xpring-eng/xpring-common-js), the server side component (TODO: link), and the [language specific client libraries](https://github.com/xpring-eng/xpring-sdk#client-side-libraries).
```
Suggest something like:

> This library is widely consumed, including by:
>
>*  [Xpring Common JavaScript](https://github.com/xpring-eng/xpring-common-js) 
> * The server side component (TODO: link) 
> * [Language specific client libraries](https://github.com/xpring-eng/xpring-sdk#client-side-libraries)

keefertaylor replied:
Done. 

Re-ordered the list, changed 'Language specific client libraries' to 'Language Specific Libraries in Xpring SDK' and dropped the TODO.

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+This library is consumed by both the [Xpring Common JavaScript](https://github.com/xpring-eng/xpring-common-js), the server side component (TODO: link), and the [language specific client libraries](https://github.com/xpring-eng/xpring-sdk#client-side-libraries).
+
+This library is widely consumed and must remain backwards compatible. Therefore, Xpring will be fairly judicious about what new functionality gets added here.
```
Perhaps add a subheading here to indicate that these two paragraphs are about change to the library. "Accepting Changes" or "Considerations for Inclusion" or even just "Backwards Compatibility"--or something better! 

Also, suggest a slight rephrase since the first sentence reiterates the previous paragraph:

> Because this library is widely consumed, it must remain backwards compatible. In most cases, Xpring will not consider adding functionality that breaks backward compatibility.

keefertaylor replied:
Done. 

I reworded a bit. It's not about breaking backwards compatibility, it is that anything that gets added here will need to be maintained in 5+ dependent libraries indefinitely.

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+This library is widely consumed and must remain backwards compatible. Therefore, Xpring will be fairly judicious about what new functionality gets added here.
+
+Additionally, code changes to this library will likely be accompanied by a feature change in one or more dependent libraries. A new pull request should document how the new fields and functionality will be used.
```
Maybe make this active, so it's addressed to contributors? i.e.:

> If you make a code change to this library, you are more than likely adding a new feature in one or more dependent libraries. Your [pull requests](#requirements-for-a-successful-pull-request) for all code changes should document how the new fields and functionality will be used.

keefertaylor replied:
Done

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+- Title is prefixed with "[WIP]" if the pull request is not ready for review.
+- Passing continuous integration.
+- Documentation in pull request about how the new functionality will be used in client libraries.
```
Here's a good template for READMEs that we could link to: https://github.com/dbader/readme-template

keefertaylor replied:
I'm confused. I mean the text of the pull request. 

In this PR it would be the blurb at the top that starts with "Update README to show contributors, link to contributions and link to top level documentation at Xpring-SDK"

Am I misunderstanding you?

ryangyoung replied:
Sorry, my comment wasn't clear. I was thinking of adding a link to the template so (hopefully) the READMEs will be at least sort of standardized, a la:

> - Documentation in pull request about how the new functionality will be used in client libraries. [Here's a template](https://github.com/dbader/readme-template) that may be useful to follow.

keefertaylor replied:
Ah. I think we're talking past each other.

I'm saying "I'd like the documentation of the PR you send me to tell me what you plan to do with this feature in detail", I am not trying to get people to update the README every time they add a new change to this library.

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+## Requirements for a Successful Pull request
+
+- Title is prefixed with "[WIP]" if the pull request is not ready for review.
+- Passing continuous integration.
```
To be consistent with other items in the list:

> - Continuous integration tests pass on PR.

keefertaylor replied:
Done (although expanded 'PR' to 'pull request'

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+- Documentation in pull request about how the new functionality will be used in client libraries.
+- Changes to protocol buffers adheres to [Google's Protocol Buffer style guide](https://developers.google.com/protocol-buffers/docs/style).
+- Changes to gRPC services adhere to [Google's API design guidelines](https://cloud.google.com/apis/design/).
+- Pull requests text and comments adhere to the [code of conduct](CODE_OF_CONDUCT.md) for this repository.
```
Suggest slight tweak here:

> - Text and comments of pull request adhere to the [code of conduct](CODE_OF_CONDUCT.md) for this repository.

ryangyoung commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+## Building The Library
+
+Any successful compilation of protocol buffers is sufficient to prove this library builds. By default, the protocol buffer comppiler  To get started:
```
typo: comppiler

Also some missing words?

keefertaylor replied:
Ooof, yeah. Sorry about that.

keefertaylor commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,76 @@
+our community a harassment-free experience for everyone, regardless of age, body
+size, disability, ethnicity, sex characteristics, gender identity and expression,
+level of experience, education, socio-economic status, nationality, personal
+appearance, race, religion, or sexual identity and orientation.
```
Totally. I didn't even notice. It's weird that they've done this on their raw markdown file: https://www.contributor-covenant.org/version/1/4/code-of-conduct.md

Modified text per your suggestion.

keefertaylor commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,76 @@
+
+## Our Standards
+
+Examples of behavior that contributes to creating a positive environment
```
Done

keefertaylor commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,76 @@
+Project maintainers have the right and responsibility to remove, edit, or
+reject comments, commits, code, wiki edits, issues, and other contributions
+that are not aligned to this Code of Conduct, or to ban temporarily or
+permanently any contributor for other behaviors that they deem inappropriate,
```
done

keefertaylor commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+# Contributing
+
+Thanks for considering a contribution to the [Xpring SDK](https://github.com/xpring-eng/xpring-sdk)!
```
Right... I remember this discussion 👍

keefertaylor commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,47 @@
+
+TODO: Diagram of Xpring SDK
+
+This library is made up of [protocol buffers](https://developers.google.com/protocol-buffers) which form common model objects for the Xpring SDK. There are also [gRPC service definitions](https://grpc.io), which form a network interace for the Xpring SDK.
```
Done

keefertaylor commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,76 @@
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported by contacting the project team at xpring@ripple.com. All
```
I'm not sure. I'm guessing the volume of mail to this address is going to be low (hopefully zero).  I am currently on this mailing list, as are other engineers who could assist. 

I can add myself as a single point of contact, if that's better?  Spinning up a whole process to handle this seems heavyweight.

ryangyoung replied:
No, I think it's fine to keep as-is. Just wanted to call out that it _may_ be something we need to change or think about more, but, yeah, hopefully there won't be any issues. Agree that it's not worth changing anything or setting up a process until we really need to.

ryangyoung commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,48 @@
+
+## Our Pledge
+
+In the interest of fostering an open and welcoming environment, we as contributors and maintainers pledge to make participation in our project and our community a harassment-free experience for everyone, regardless of, but not limited to characteristics, age, body size, disability, ethnicity, sex characteristics, gender identity and expression, level of experience, education, socio-economic statu ...
```
Sorry, my suggested change had a flaw. I think that should be:

> regardless of, but not limited to characteristics **like** age, body size, etc etc

keefertaylor replied:
Done

ryangyoung commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,48 @@
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team at [xpring@ripple.com](mailto:xpring@ripple.com). All complaints will be reviewed and investigated and will result in a response that is deemed necessary and appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an  ...
```
> [xpring@ripple.com](mailto:xpring@ripple.com)

FWIW, that works, but in Markdown, you can also use angle brackets for email addresses and URLs, like so:

```
<xpring@ripple.com>
```

which resolves to <xpring@ripple.com>

keefertaylor replied:
Done. Thanks for schooling me.

keefertaylor commented:
Thank you for the detailed comments. I have attached a diagram that I put together. 

I have an open request to design to make something a bit more on-brand and more approachable tracked in Asana.

vaelinn123 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,53 @@
+
+## Requirements for a Successful Pull request
+
+- Title is prefixed with "[WIP]" if the pull request is not ready for review.
```
Just throwing out a bit to think about, instead of a "[WIP]" prefix on pull requests you could use [draft](https://github.blog/2019-02-14-introducing-draft-pull-requests/) pull requests, these can be filtered out when looking at the pending reviews with `draft:false` in the github PR search. For example, this is my standard search when looking for PR's ready for review: `is:pr is:open review:none draft:false -author:vaelinn123` 

Just throwing out ideas, adding the prefix is probably more intuitive.

intelliot commented:
I'm not too sure about `architecture.png`; what's the meaning of the arrows (and arrow direction)? Also, rippled seems to be in the first row. I might expect rippled to be at the lower right (next to Common Core), or maybe on a third row of its own.

keefertaylor commented:
Re Diagram:
In the architecture diagram, items at the top live in client land, making them things a user may want to actually use. 

The bottom row is the core / internal libraries which power the end. The arrows represent dependencies between components, akin to how we’d represent classes here as a UML diagram.

keefertaylor commented:
Vaelinn: Thanks for the feedback and helping us out!

Re Linter: I didn’t realize protocol buffer linters were a thing. I’ll check this out tomorrow. Most of our other libraries (save Java) have a linter in place. My intention is to get linting set up for everything. 


Let me check out how this work flow works tomorrow (I’ve never used draft PRs before) but this sounds like a much better way than my special prefixing proposal.

vaelinn123 commented:
I may have misspoke regarding the linter here, I was more referencing a documentation section that informs a contributor on how to verify that their code changes would properly pass the linting mentioned in your response [here](https://github.com/xpring-eng/Xpring-JS/issues/22#issuecomment-537729880). 

May not be applicable here.

keefertaylor commented:
I've gone ahead and wired a linter here: 
https://github.com/xpring-eng/xpring-common-protocol-buffers/pull/11

Updated documentation on PR requirements to reflect this, and updated the building instructions as well. 

LMK if anyone has outstanding concerns.

keefertaylor commented:
That is correct.
